_posts/2016-03-16-code-smells-part-I.md

```diff
@@ -0,0 +1,259 @@
+    return post;
+}
+
+````
+
+Then we can do it for the parameters of the request:
+
+````
+- (NSArray *)createPostParams:(NSString *)post
+{
+    NSMutableDictionary *postParams = [NSMutableDictionary dictionary];
+    [postParams setObject:txtTitle.text forKey:@"title"];
+    [postParams setObject:post forKey:@"data"];
+    [postParams setObject:txtUser.text forKey:@"username"];
+    [postParams setObject:txtPassword.text forKey:@"password"];
+
+    NSArray *args = [NSArray arrayWithObjects:[NSNumber numberWithInt:0], postParams, nil];
+    return args;
+}
+
+````
+
+With all this in place we are now ready to create a XMLRPCRequest:
+
+````
+- (XMLRPCRequest *)createXMLRPCRequestWithArgs:(NSArray*)args {
+
+    XMLRPCRequest *request;
+
+#ifdef DEBUG_LOG
+   request = [[XMLRPCRequest alloc] initWithURL:
+                              [NSURL URLWithString:@"http://localhost:8888/letsbasket/xmlrpc.php"]];
+    DLog(@"Debug");
+#else
+    request = [[XMLRPCRequest alloc] initWithURL:[NSURL URLWithString:[UtilsHelper localizeString:@"UrlXmlRPCKey"]]];
+    DLog(@"Producao");
+#endif
+
+    [request setMethod:@"letsBasket.AddPost" withParameters:args];
+
+    return request;
+}
+
+````
+We can also extract a method with some display updates:
+
+````
+- (void)updateDisplay
+{
+    UIApplication *app = [UIApplication sharedApplication];
+    app.networkActivityIndicatorVisible = NO;
+
+    [self dismissWaitingAlert];
+}
+````
+And last but not least we can extract the preparation for displaying the error message:
+
+````
+- (void)showError:(NSString*)bodyResult {
+
+    int location_start = [bodyResult rangeOfString:@"<string>"].location + 8;
+    int location_end = [bodyResult rangeOfString:@"</string>"].location;
+
+    NSString *message = [[bodyResult substringWithRange:NSMakeRange(location_start, location_end- location_start)] unescapedString];
+    NSString* title = [UtilsHelper localizeString:@"PublishVC_ErrorRetreivingAlertTitle_key"];
+    [self showAlertWithErrorMessage:message Title:title];
+}
+````
+With all these extractions our method now looks pretty neat:
+
+````
+- (void) postToServer
+{
+    NSString *post = [self serializePost];
+    NSArray *args = [self createPostParams:post];
+    XMLRPCRequest *request = [self createXMLRPCRequestWithArgs: args];
+    NSError *error = nil;
+
+	XMLRPCResponse* result = [XMLRPCConnection sendSynchronousXMLRPCRequest:request error:&error];
+
+    [self updateDisplay];
+
+    if(error != nil || [[result body] rangeOfString:@"<name>error</name>"].location != NSNotFound)
+    {
+        [self showError:[result body]];
+		return;
+	}
+
+    [self processPublishResult:result];
+}
+````
+
+Hum... we can do this even better... let's look at the method createXMLRCPRequest and see if we can call the others from there...in this case, it makes sense to have all together.
```

Comment:
Hum... we can do this even better! Let's take a look at the method `createXMLRCPRequest` and see if we can call the others from there. In this case, it makes sense to have all together.